module/mempool/stdmap/backdata/heropool/pool.go

```diff
@@ -0,0 +1,300 @@
+// the index at which given entity is written on entities linked-list so that it can be accessed directly later.
+func (e *Pool) Add(entityId flow.Identifier, entity flow.Entity, owner uint64) EIndex {
+	entityIndex := e.sliceIndexForEntity()
+	e.values[entityIndex].entity = entity
+	e.values[entityIndex].id = entityId
+	e.values[entityIndex].owner = owner
+	e.values[entityIndex].node.next.setUndefined()
+	e.values[entityIndex].node.prev.setUndefined()
+
+	if e.used.head.isUndefined() {
+		// used list is empty, hence setting head of used list to current entityIndex.
+		e.used.head.setPoolIndex(entityIndex)
+		e.values[e.used.head.sliceIndex()].node.prev.setUndefined()
+	}
+
+	if !e.used.tail.isUndefined() {
+		// links new entity to the tail
+		e.connect(e.used.tail, entityIndex)
+	}
+
+	// since we are appending to the used list, entityIndex also acts as tail of the list.
+	e.used.tail.setPoolIndex(entityIndex)
+
+	e.size++
+	return entityIndex
+}
+
+// Get returns entity corresponding to the entity index from the underlying list.
+func (e Pool) Get(entityIndex EIndex) (flow.Identifier, flow.Entity, uint64) {
+	return e.values[entityIndex].id, e.values[entityIndex].entity, e.values[entityIndex].owner
+}
+
+// sliceIndexForEntity returns a slice index which hosts the next entity to be added to the list.
+func (e *Pool) sliceIndexForEntity() EIndex {
+	if e.free.head.isUndefined() {
+		// the free list is empty, so we are out of space, and we need to eject.
+		if e.ejectionMode == RandomEjection {
+			// turning a random entity into a free head.
+			e.invalidateRandomEntity()
+		} else {
+			// LRU ejection
+			// the used head is the oldest entity, so we turn the used head to a free head here.
+			e.invalidateUsedHead()
+		}
+	}
+
+	// claiming the head of free list as the slice index for the next entity to be added
+	return e.claimFreeHead()
+}
+
+// Size returns total number of entities that this list maintains.
+func (e Pool) Size() uint32 {
+	return e.size
+}
+
+// getHeads returns entities corresponding to the used and free heads.
+func (e Pool) getHeads() (*cachedEntity, *cachedEntity) {
+	var usedHead, freeHead *cachedEntity
+	if !e.used.head.isUndefined() {
+		usedHead = &e.values[e.used.head.sliceIndex()]
+	}
+
+	if !e.free.head.isUndefined() {
+		freeHead = &e.values[e.free.head.sliceIndex()]
+	}
+
+	return usedHead, freeHead
+}
+
+// getTails returns entities corresponding to the used and free tails.
+func (e Pool) getTails() (*cachedEntity, *cachedEntity) {
+	var usedTail, freeTail *cachedEntity
+	if !e.used.tail.isUndefined() {
+		usedTail = &e.values[e.used.tail.sliceIndex()]
+	}
+
+	if !e.free.tail.isUndefined() {
+		freeTail = &e.values[e.free.tail.sliceIndex()]
+	}
+
+	return usedTail, freeTail
+}
+
+// connect links the prev and next nodes as the adjacent nodes in the double-linked list.
+func (e *Pool) connect(prev poolIndex, next EIndex) {
```

Comment:
This seems more complicated than it is because `connect(prev poolIndex, next EIndex)` would look simpler as `connect(prev poolIndex, next poolIndex)`. Another comment suggests refactoring `poolIndex` and `EIndex` to be the same...

Reply:
`connect` method is always called with second parameter an `EIndex`. Refactoring it to `connect(prev poolIndex, next poolIndex)` requires the caller to convert `poolIndex` to `EIndex`, which makes the code less DRY (Don't Repeat Yourself).

Reply:
> requires the caller to convert poolIndex to EIndex

The intention was / suggestion would be to refactor `EIndex` away at some point so that only `poolIndex` exists, and therefore the caller would not need to convert anything :-)

`poolIndex` and `EIndex` are nearly the same and ironically already making the code DRY? :-) One of the comments already says "getSliceIndex returns the slice-index equivalent of the poolIndex".

The two types of index only seem to exist because `poolIndex` `0` has a special meaning. If the code biz logic would be refactored so that the special meaning of `poolIndex` zero was achieved / implemented differently, then `poolIndex` and `EIndex` could be exactly the same -- i.e. both starting from zero and meaning the same thing -- and thus `EIndex` could disappear and `poolIndex` would be used everywhere where `poolIndex` and `EIndex` are currently used? And no conversion by the caller either... :-)

Reply:
This comment seems a duplication of an [earlier conversation we had](https://github.com/onflow/flow-go/pull/1778#discussion_r779081909)